ext/bonito/connection.jl

```diff
@@ -0,0 +1,164 @@
+using HTTP.WebSockets: WebSocket, WebSocketError
+using HTTP.WebSockets: receive, isclosed
+using HTTP.WebSockets
+
+using Oxygen: Oxygen
+using Bonito: Bonito
+using Bonito: FrontendConnection, WebSocketHandler, Session, setup_websocket_connection_js
+using Bonito: run_connection_loop, register_connection!
+using Bonito: CleanupPolicy, DefaultCleanupPolicy, allow_soft_close, soft_close, should_cleanup
+
+mutable struct OxygenWebSocketConnection <: FrontendConnection
+    context::Oxygen.ServerContext
+    endpoint::String
+    session::Union{Nothing,Session}
+    handler::WebSocketHandler
+end
+
+function OxygenWebSocketConnection(context::Oxygen.ServerContext, endpoint::String)
+    return OxygenWebSocketConnection(context, endpoint, nothing, WebSocketHandler())
+end
+
+Base.isopen(ws::OxygenWebSocketConnection) = isopen(ws.handler)
+Base.write(ws::OxygenWebSocketConnection, binary) = write(ws.handler, binary)
+Base.close(ws::OxygenWebSocketConnection) = close(ws.handler)
+
+mutable struct BonitoConnectionContext
+    cleanup_policy::CleanupPolicy
+    open_connections::Dict{String, OxygenWebSocketConnection}
+    cleanup_task::Union{Task, Nothing}
+    lock::ReentrantLock
+end
+
+BonitoConnectionContext(policy=DefaultCleanupPolicy()) = BonitoConnectionContext(policy, Dict{String, Session{OxygenWebSocketConnection}}(), nothing, ReentrantLock())
+
+"""
+    setup_bonito_connection(
+        context::Context;
+        setup_all=false,
+        setup_route=setup_all,
+        setup_register_connection=setup_all,
+        route_base="/bonito_websocket/"
+    )
+
+This is the high level function to setup bonito connection. It will register the route and create a connection if needed.
+
+In the simplest use case you can simply pass `setup_bonito_connection(CONTEXT[]; setup_all=true)` and it will set up everything for you. Please see the guide for advanced usage.
+"""
+function Oxygen.setup_bonito_connection(
+    context::Oxygen.ServerContext;
```

Comment:
Thanks for adding the unit tests, that gives me some peace of mind and also helps me understand how to use this extension. 

Now about this function here - let's brainstorm here a bit. It wouldn't be ideal to have the public api to have to explicitly use the `CONTEXT[]` object, I'd like to keep that part purely internal. 

We might be able to follow the pattern used inside the methods.jl file where we bind the server context to the stateful methods inside core. We could try placing a public facing verions of this function inside the exts.jl file and have it point to this internal one. But this is just an idea, i'm sure there are other ways to do this and I'd like to hear what you think

Reply:
Yeah it is a wrinkle. This only works with `methods.jl` because it's included both globally and with `@oxydise`, I think. As you say it would be possible to add an overloaded version to `methods.jl`, this is quite straightforward, if slightly unfortunate to have to spread code from extensions everywhere. Also we would seem to have a method even when the extension is not loaded, but that's just how Julia extensions work I suppose. The only other way I can think of is to use a macro, which could just grab `CONTEXT` from the environment, but macros are meant to be used sparingly and are kind of a pain to write, so I'd probably err towards the first if you think it's okay to put more stuff in `methods.jl`.

Reply:
I think placing it in the methods could work, I'd also be open to having an "ext_methods.jl" module just for stateful extension methods so we can keep these specific use cases grouped together. The regular `ext.jl` just holds the partial definitions but none of the functions are bound to the server context.

We'd have to load it in both places just like the methods.jl file. What do you think of this approach?

Reply:
Ok seems fine to me -- done. The test failures seem to be caused a problem with Coveralls' servers.

Reply:
No worries, they've been down for maintenance today. As long as the tests pass we know the coverage is good since they completed the other day